### PR TITLE
Store the typedef identifier in the symbol_type

### PR DESCRIPTION
**Pull request by thk123**
Issue: https://github.com/diffblue/verification-engine-utils/issues/315

_This is WIP and should not be merged_

The idea here is to, in the front end, store the typedef identifier into the symbolic type. Then, in `expr2ct` when writing a symbolic type, we just use the identifier stored in the `ID_typedef` instead of trying to convert the concrete type. This way, anon structs etc get a proper type. 

There is plenty of obvious tidying to do. I am opening this PR to get feedback on:

1. Is this approach reasonable
2. Where do the symbolic type symbols come and go. I seem to add ID_typedef to a symbolic type, but its type symbol in the symbol_table is removed in `remove_internal_symbols`. However, by the time the `expr2ct` comes to convert a type that uses the typedef, the type is a symbolic type.

### CODE REVIEW

**Comment by thk123** [comment]
Chat with Peter about this:

> thk123
> I've found a place where I can move the `ID_typedef` tag onto the real type, but what happens if more than one typedef for each given type. However, I think I agree with previous comment about  the real type being the only space place to definitely find it so perhaps can have a list of them under the same ID. (Think because if an expression as a symbolic type as its type then isn't it exactly that symbolic name we wish to use (e.g. if I typedef int as MYINT, I want it to use MYINT not int even if I can follow it?
> 
> peterschrammel   
> But how would we decide which typedef to use when generating the code?
> It would be rather an aesthetic problem than a functional one because it would compile with whatever type we choose.
> What about chains of typedefs?
> typedef int myint; typedef myint myint2; ....
> 
> thk123  
> Re: multiple typedefs using the wrong name if we store all the types: We have the problem of potentially using the wrong typedef which ever since if we only store one and it is typedef'd to two different names and we only store one then the name will be wrong half the time. Chains would also work in the same way, when processing `typedef myint myint2;` we'd follow the chain down and add (or overwrite) the `ID_typedef` in the int type. Though this would be solved by _not_ following symbolic types when generating code and instead using them directly (this would retain the correct typedef'd version of the name and let the regular C compiler handle resolving the types  
> At the moment, expr2c, when converting a type with ID `ID_symbol` instead finds the concrete type and converts that. But if it just doesn't and instead uses its `ID_identifier` then providing the typedef is still included (which we need anyway to compile I think?) this will be a valid generated code.
> I've now talked myself around to this I think - we don't care what the "real" type is when we are declaring the variable, and when we are assigning it, we don't care what the type name is
> 
> peterschrammel
> There is no concept of type aliases in goto programs.
> That's why we have to find our way around it. Somehow.
> 
> thk123  
> I'm not sure I see the problem. The GOTO program works correctly? The problem is only in generating C code at the end - surely at this point since we know the parameter is of type `example_s` (a symbolic type) we can just use that directly
> 
> peterschrammel
> The problem is when in the goto-program you don't have example_s in certain places but the type that example_s defines. Then we have to revert it back to example_s. If this link is not invertible then we can only pick one of the possible choices.
> 
> thk123 
> But the change I am proposing is strictly after the goto program is done. I am proposing the expr2c (it could even be expr2cleanc to really isolate the change) when producing C for a symbolic type, use the identifier rather than produce C for the underlying real type. What am I missing? Are you suggesting that maybe we'd get to the C test code generation step and the type of the parameter would not be the symbolic type but the concrete type?
> 
> peterschrammel
> I'm not sure but this could happen. But let's let it happen first and think about it then.

TL;DR - there may be a problem where the typedef is lost completely, but since couldn't think of when will ignore. This solution at least allows us to use the right typedef in each case even in the event of chains or multiple typedef's on the same type.

**Comment by tautschnig** [comment]
The closest analogy I could see is ID_C_cformat, where we retain the original expression with constants. I'm fairly sure it should be ID_C_typedef to make explicit that this does not affect semantics. Then, store it with the expression every time a typedef is expanded.

**Comment by thk123** [comment]
@tautschnig I don't really understand, aside from replacing `ID_typedef` with `ID_C_typedef` is this different to the proposed change? I guess it is the storing it in every expression where the typedef is expanded, but I'm not sure where that is?

**Comment by tautschnig** [comment]
It might not actually not be different at all from what you've been doing already. What I would expect to find is that any expression the type of which is an expanded typedef has the structure
```
symbol:
* type:
  unsignedbv
  * width
    32
  * #source_location
    ...
  * #typedef
    int32_t
* identifier
  some_name
```

File: src/ansi-c/expr2c.cpp [hunk 1]
```diff
@@ -378,7 +383,10 @@ std::string expr2ct::convert_rec(
 
     dest+=" }";
 
-    dest+=d;
+    if(src.find(ID_typedef).is_not_nil())
+      dest+=src.get_string(ID_typedef);
+    else
+      dest+=d;
```
Comment (thk123) [inline src/ansi-c/expr2c.cpp]: @peterschrammel How come we only add on the `d` (declarator) string if not a typedef. Not really clear on exactly what could be in `d` at this point.

**Comment by thk123** [comment]
Finally getting my head around the problem. As I understand it, what happens is:

1. when parsing the C program we create symbolic types for each typedef (Here?: https://github.com/diffblue/cbmc/pull/345/files#diff-ea8c3a90069cd357990d58399df18ec4R142?)
2. then when converting into a goto program we throw away all these symbolic types * and attach the ID_typedef to the concrete type **
3. When generating the code, we can use the concrete type and identify the typedef'd name. 

(*)Somehow these symbolic types remain if you use a .c file directly, but are lost if you compile with goto-cc
(**) This will only work as expected if at most one typedef for a given type. If chained or duplicated, this information will be lost. 

@tautschnig I'm a bit confused about where your structure is. Is it on the symbol for a given variable? If so, is the type a reference to the concrete type? In this case, how do you envisage dealing with chained/multiple typedefs? It seems the typedef should actually be attached to the variable symbol which typedef it is using? 

I would have guessed this can be added in the same place, but by that point the type has the mysterious (to me!) value "already-typechecked". Instead perhaps it needs to go earlier, e.g. in the parsing `ansi_c_parsert::add_declarator`?

File: src/linking/remove_internal_symbols.cpp [hunk 2]
```diff
@@ -144,6 +144,16 @@ void remove_internal_symbols(
     if(is_type)
     {
       // never EXPORTED by itself
+        // Without this code the symbol is lost when compiling with
```
Comment (thk123) [inline src/linking/remove_internal_symbols.cpp]: Here is where I _was_ copying the ID_typedef into the concrete type (without this [or the above change]) then we don't have the typedef'd name of the type. However, as described in the comment, this won't work with chained or multiple typedefs.

**Comment by thk123** [comment]
See my latest commit for a proof of concept of I think how to assign the ID_typedef to each symbol rather than to the underlying type only (which should allow accurate recreation of the correct typedef). 

It doesn't yet work as the interpreter in `load_counter_example_inputs` is for some reason not giving the right type, but the info does propogate as in the symbol table all the relevant entries have a typedef with appropriate value even using goto-cc and having two defined names for a type. 

Thoughts?

**Comment by thk123** [comment]
So in doing some more rigorous testing I have thoroughly confused myself. I thought I needed to _either_ copy the symbol when we remove it, or somehow store it in the parsing step. However, without either of them, all I need to get it working is the modification in `expr2ct` that if the type has a `ID_C_typedef` flag, we use that rather than the concrete type. 

I have an outstanding question on exactly how the `expr2ct::convert_rec` should behave, but I will create a clean PR with the minimal changes and the tests that demonstrate the behaviour and discuss it there.

File: src/ansi-c/expr2c.cpp (hunk 1 above)
Comment (thk123) [inline src/ansi-c/expr2c.cpp]: This does appear to be a bug, for the following code:

    typedef struct tag_struct_name
    {
    	int x;
    	float y;
    } MYSTRUCT;
    
    void fun()
    {
    	   const struct tag_struct_name tag_struct_var = {.x = 1, .y = 3.14f};
    	   const MYSTRUCT mystruct_var = {.x = 3, .y = 2.1f};
    }

I get the following symbols:

    Symbol......: fun::1::mystruct_var
    Pretty name.: fun::1::mystruct_var
    Module......: main
    Base name...: mystruct_var
    Mode........: C
    Type........: const MYSTRUCT
    Value.......: { .x=3, .y=2.100000e+0f }
    Flags.......: lvalue thread_local file_local
    Location....: file main.c line 11 function fun
    
    Symbol......: fun::1::tag_struct_var
    Pretty name.: fun::1::tag_struct_var
    Module......: main
    Base name...: tag_struct_var
    Mode........: C
    Type........: struct tag_struct_name
    Value.......: { .x=1, .y=3.140000e+0f }
    Flags.......: lvalue thread_local file_local
    Location....: file main.c line 10 function fun

Comment (thk123) [inline src/ansi-c/expr2c.cpp]: So this seems unrelated as I get this problem on the main branch, I have created an issue #355 and a PR that includes a regression test checking for it: #356

**Comment by thk123** [comment]
As promised, here is the cleaned up, minimal version of this change with extended tests: https://github.com/diffblue/cbmc/pull/358
